### PR TITLE
Updating HfApi objects to use dataclass

### PR DESCRIPTION
**Pull request by Ahmedniz1**
This PR resolves issue #1911 
I've updated all classes to use dataclass except (transformerinfo class).

### CODE REVIEW

File: src/huggingface_hub/hf_api.py [hunk 1]
```diff
@@ -238,11 +238,15 @@ def repo_type_and_id_from_hf_id(hf_id: str, hub_url: Optional[str] = None) -> Tu
...
 
 
-class LastCommitInfo(TypedDict, total=False):
+@dataclass
+class LastCommitInfo(dict):
     oid: str
     title: str
     date: datetime
 
+    def __post_init__(self):  # hack to make BlobLfsInfo backward compatible
```
Comment (Wauplin) [inline src/huggingface_hub/hf_api.py:247]:
```suggestion
    def __post_init__(self):  # hack to make LastCommitInfo backward compatible
```

File: src/huggingface_hub/hf_api.py [hunk 2]
```diff
@@ -254,11 +258,15 @@ def __post_init__(self):  # hack to make BlobLfsInfo backward compatible
...
 
 
-class BlobSecurityInfo(TypedDict, total=False):
+@dataclass
+class BlobSecurityInfo(dict):
     safe: bool
     av_scan: Optional[Dict]
     pickle_import_scan: Optional[Dict]
 
+    def __post_init__(self):  # hack to make BlobLfsInfo backward compatible
```
Comment (Wauplin) [inline src/huggingface_hub/hf_api.py:267]:
```suggestion
    def __post_init__(self):  # hack to make BlobSecurityInfo backward compatible
```

File: src/huggingface_hub/hf_api.py [hunk 3]
```diff
@@ -268,10 +276,14 @@ class TransformersInfo(TypedDict, total=False):
     processor: Optional[str]
 
 
-class SafeTensorsInfo(TypedDict, total=False):
+@dataclass
+class SafeTensorsInfo(dict):
     parameters: List[Dict[str, int]]
     total: int
 
+    def __post_init__(self):  # hack to make BlobLfsInfo backward compatible
```
Comment (Wauplin) [inline src/huggingface_hub/hf_api.py:284]:
```suggestion
    def __post_init__(self):  # hack to make SafeTensorsInfo backward compatible
```

**Review by Wauplin — APPROVED** [review]
Thanks @Ahmedniz1! 
Left 3 small comments to update but that's all. Then we should be able to merge :tada:

**Comment by Ahmedniz1** [comment]
@Wauplin I've updated the comments. 
Thanks

**Pull request merged**